Move to ESM

tommy-mitchell opened the pull request:
Closes #183.

Checklist:

- [X] Update `tsconfig.tsd.json`
	* Set `module` to `node16`
	* Set `target` to `ES2020`
- [X] Update `package.json`
	* Add `"type": "module"`
	* Update to `AVA` v5.2.0, add [`tsx`](https://github.com/esbuild-kit/tsx) for transpilation
		- [ ] Do we need to add "type checking" command? https://github.com/esbuild-kit/tsx#does-it-do-type-checking
	* Inline `clean` script
- [X] Add `.js` to relative imports in source
	* Is this needed for barrel imports as well? E.g:
		```ts
		import tsd from './lib';
		// vs
		import tsd from './lib/index.js';
		```
- [ ] Add `.js` to relative imports in test
	* Is this necessary either? I believe `tsx` supports "Node"-style imports within ESM.
	* Related, with `tsx` should `execa` expect `cli.js` or `cli.ts`?
- [ ] Make sure that Node builtins use the `node:` protocol

Speaking of tests, trying to run `AVA` causes errors with `@tsd/typescript`:

```sh
npx ava source/test/test.ts

(node:69218) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
  Uncaught exception in source/test/test.ts

  SyntaxError: The requested module '@tsd/typescript' does not provide an export named 'TypeFormatFlags'

  ✘ source/test/test.ts exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

In addition, updating `AVA` means that `t.throwsAsync` can also return `undefined`:

```ts
test('failure', async t => {
	const cwd = path.join(__dirname, 'fixtures/failure');
	const {exitCode} = await t.throwsAsync<ExecaError>(execa('../../../cli.js', {cwd}));
	//=> Property 'exitCode' does not exist on type 'ExecaError | undefined'.

	t.is(exitCode, 1);
});
```

---

cc: @sindresorhus

tommy-mitchell commented:
Using `tsx` would resolve #175 as well.

mrazauskas commented:
By the way, this is 100% breaking change for programmatic API users. Perhaps #75 could be considered? Would be enough to return an object. This would make it easier to add any additional props in the future.

The #75 was needed for [`jest-runner-tsd`](https://github.com/jest-community/jest-runner-tsd), but it was rather unclear if that will ever happen. I fork `tsd` and released [`tsd-lite`](https://github.com/mrazauskas/tsd-lite). You might be interested to look around:

- [@tsd/typescript](https://npmjs.com/package/@tsd/typescript) package is moved to peer dependencies to allow testing on different versions of TS;
- works with monorepos;
- has less bugs. For instance, #174 or #141 is not an issue.

sindresorhus commented:
What's the benefit of using `.tsx`? It's not clear to me.

sindresorhus commented:
I think what you want is this: https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#allowimportingtsextensions

sindresorhus commented:
> SyntaxError: The requested module '@tsd/typescript' does not provide an export named 'TypeFormatFlags'

The model resolution is stricter about exports. Make sure it's actually exported correctly.

tommy-mitchell commented:
> What's the benefit of using `.tsx`? It's not clear to me.

`tsx` is a tool like `ts-node` for transpiling TS in-memory. The name comes from modeling `npx`. It’s faster than `ts-node` and works better with ESM, but skips type checking.

The naming is unfortunate.

https://github.com/avajs/ava/issues/2593#issuecomment-1345069993

sindresorhus commented:
> tsx is a tool like ts-node for transpiling TS in-memory. The name comes from modeling npx. It’s faster than ts-node and works better with ESM, but skips type checking.

Ok. We definitely want type checking to ensure they are valid. But we could also just run `tsc --noEmit` before to type check.

sindresorhus commented:
Let me know if anything is blocked on me commenting on something.

tommy-mitchell commented:
> But we could also just run `tsc --noEmit` before to type check.

I decided to go with that as well. I've been doing more locally, you're fine.

> SyntaxError: The requested module '@tsd/typescript' does not provide an export named 'TypeFormatFlags'

`TypeFormatFlags` is an enum exported by`typescript` itself. I'm going to see if using `ts-node` works instead.

---

Re: #75, should we close that PR and open a new one? It's definitely stale, and the diff is pretty extraneous at this point. I can open a reminder issue.

tommy-mitchell commented:
Seems like the issue with the tests is that they are being transpiled to CJS, but since `"type": "module"` is set the tests throw.

Related, is there a reason that we have `tsconfig.tsd.json` and not `tsconfig.json`?

sindresorhus commented:
> Related, is there a reason that we have tsconfig.tsd.json and not tsconfig.json?

https://github.com/SamVerschueren/tsd/pull/100

tommy-mitchell commented:
I've uploaded some more changes (tests are still failing). I'll resolve the merge conflicts.

A couple highlights: moved to `xo` for linting, updated the `test` script, and consolidated the tests via macros (still need to do so for the CLI tests).

tommy-mitchell commented:
> > SyntaxError: The requested module '@tsd/typescript' does not provide an export named 'TypeFormatFlags'
> 
> The ES module resolution is stricter about exports. Make sure it's actually exported correctly.

Turns out this is because the TypeScript compiler is CJS: https://github.com/microsoft/TypeScript/issues/32949

tommy-mitchell commented on source/test/cli.ts:
```diff
@@ -1,123 +1,122 @@
-	t.true(formattedResults.includes('✖  5:19  Argument of type number is not assignable to parameter of type string.'));
+	const options = {cwd: getFixture('failure')};
+
+	const {stderr: cliResults} = await t.throwsAsync<ExecaError>(execa(binPath, options));
```
What's the best way to test that `execa` throws while still being able to easily test the result? Since with `AVA` 4+, the result of `t.throws*` can be `undefined`.

sindresorhus replied:
Either guard it or force unwrap:

```suggestion
	const {stderr: cliResults} = await t.throwsAsync<ExecaError>(execa(binPath, options))!;
```

sindresorhus commented:
@tommy-mitchell Just a reminder that this is still open. If you're simply busy, feel free to ignore this.

tommy-mitchell commented:
@sindresorhus this is on my todo list, I’ve got some local changes that are unpushed. If you have some time, I’d love your feedback on #196

sindresorhus commented on source/lib/config.ts:
```diff
@@ -1,81 +1,73 @@
 	};
 
-	const module = combinedCompilerOptions.module ?? ModuleKind.CommonJS;
+	const module = combinedCompilerOptions.module ?? ts.ModuleKind.CommonJS;
```
```suggestion
	const module = combinedCompilerOptions.module ?? (pkg.type === 'module' ? ts.ModuleKind.Node16 : ts.ModuleKind.CommonJS);
```

?

tommy-mitchell replied:
Thoughts on these defaults?

```ts
function getModuleResolution(module: ts.ModuleKind): ts.ModuleResolutionKind {
	if (module === ts.ModuleKind.NodeNext) {
		return ts.ModuleResolutionKind.NodeNext;
	}

	if (module === ts.ModuleKind.Node16) {
		return ts.ModuleResolutionKind.Node16;
	}

	if (module === ts.ModuleKind.ES2020 || module === ts.ModuleKind.ES2022 || module === ts.ModuleKind.ESNext) {
		return ts.ModuleResolutionKind.Bundler;
	}

	// Fallback to Node16
	return ts.ModuleResolutionKind.Node16;
}
```

tommy-mitchell commented:
Ended up removing the `index.d.ts` requirement in this PR because it was giving me trouble with tests. Merged the changes to main. Next steps:

- Update `tsd`'s defaults
- Update PR description (e.g. this also includes a change the build script, added `@sindresorhus/tsconfig`)
- Fix test failures (CLI tests in particular don't seem to be working)
- Update documentation to reflect ESM / `index.d.ts` changes (more doc improvements can be saved for another PR)

tommy-mitchell commented:
Are we still supporting testing types in CJS packages?

tommy-mitchell commented:
Seems like `xo` is failing because it can't find a `tsconfig.json`, even though I set it in `parserOptions.project`:

```
file:///home/runner/work/tsd/tsd/node_modules/xo/lib/options-manager.js:163
		options.tsConfig = tsConfigResolvePaths(getTsconfig(options.tsConfigPath).config, options.tsConfigPath);
		                                                                         ^

TypeError: Cannot read properties of null (reading 'config')
    at handleTSConfig (file:///home/runner/work/tsd/tsd/node_modules/xo/lib/options-manager.js:163:76)
    at mergeWithFileConfig (file:///home/runner/work/tsd/tsd/node_modules/xo/lib/options-manager.js:137:19)
    at async parseOptions (file:///home/runner/work/tsd/tsd/node_modules/xo/lib/options-manager.js:578:51)
    at async Promise.all (index 0)
    at async getOptionGroups (file:///home/runner/work/tsd/tsd/node_modules/xo/lib/options-manager.js:594:21)
    at async Object.lintFiles (file:///home/runner/work/tsd/tsd/node_modules/xo/index.js:80:17)
    at async file:///home/runner/work/tsd/tsd/node_modules/xo/cli.js:212:18

```

tommy-mitchell commented:
Seems like there should be a way for us to just `tsconfig.json` directly, I'll try to see if I can do that without running into the issues from #100.

sindresorhus commented:
> Are we still supporting testing types in CJS packages after moving to ESM?

No
